pkg/tlsx/openssl/openssl.go

```diff
@@ -0,0 +1,239 @@
+//go:build (linux || darwin) && openssl
+
+// Package openssl implements a tls grabbing implementation using openssl
+package openssl
+
+import (
+	"encoding/pem"
+	"io/ioutil"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/zmap/zcrypto/x509"
+
+	"github.com/pkg/errors"
+
+	"github.com/projectdiscovery/fastdialer/fastdialer"
+	"github.com/projectdiscovery/iputil"
+	"github.com/projectdiscovery/tlsx/pkg/tlsx/clients"
+	"github.com/spacemonkeygo/openssl"
+
+	zasn1 "github.com/zmap/zcrypto/encoding/asn1"
+	zpkix "github.com/zmap/zcrypto/x509/pkix"
+)
+
+// Client is a TLS grabbing client using crypto/tls
+type Client struct {
+	dialer           *fastdialer.Dialer
+	openSSLDialFlags []openssl.DialFlags
+	options          *clients.Options
+}
+
+// New creates a new grabbing client using crypto/tls
+func New(options *clients.Options) (*Client, error) {
+	c := &Client{
+		dialer:  options.Fastdialer,
+		options: options,
+	}
+
+	if !options.VerifyServerCertificate {
+		c.openSSLDialFlags = append(c.openSSLDialFlags, openssl.InsecureSkipHostVerification)
+	}
+
+	return c, nil
+}
+
+func (c *Client) OpenSSLDialFlags() openssl.DialFlags {
+	if len(c.openSSLDialFlags) == 0 {
+		return 0
+	}
+	initialFlag := c.openSSLDialFlags[0]
+	allDialFlags := initialFlag
+	for _, dialFlag := range c.openSSLDialFlags {
+		if dialFlag != initialFlag {
+			allDialFlags = allDialFlags & dialFlag
+		}
+	}
+	return allDialFlags
+}
+
+// Connect connects to a host and grabs the response data
+func (c *Client) ConnectWithOptions(hostname, ip, port string, options clients.ConnectOptions) (*clients.Response, error) {
+	address := net.JoinHostPort(hostname, port)
+
+	if c.options.ScanAllIPs || len(c.options.IPVersion) > 0 {
+		address = net.JoinHostPort(ip, port)
+	}
+
+	opensslCtx, err := openssl.NewCtx()
+	if err != nil {
+		return nil, err
+	}
+
+	if c.options.Timeout > 0 {
+		opensslCtx.SetTimeout(time.Duration(c.options.Timeout) * time.Second)
+	}
+
+	if len(c.options.Ciphers) > 0 {
+		if err := opensslCtx.SetCipherList(strings.Join(c.options.Ciphers, ",")); err != nil {
+			return nil, errors.Wrap(err, "could not set ciphers")
+		}
+	}
+
+	if c.options.CACertificate != "" {
+		caCert, err := ioutil.ReadFile(c.options.CACertificate)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read ca certificate")
+		}
+		caStore := opensslCtx.GetCertificateStore()
+		err = caStore.LoadCertificatesFromPEM(caCert)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not add certificate to store")
+		}
+	}
+
+	conn, err := openssl.Dial("tcp", address, opensslCtx, c.OpenSSLDialFlags())
```

Comment:
We should also set openssl verify mode to None to disable errors to allow more invalid certs.

Example of an error with current implementation occuring during cert verification. Ideally we should do as lax parsing as possible. The same host works with openssl cli client which does zero validation by default.

```
[WRN] Could not connect input 24.196.75.118:443: could not connect to host: SSL errors: SSL routines:ssl3_read_bytes:sslv3 alert bad certificate
```

Reply:
done with raw connections and ignoring handshake errors